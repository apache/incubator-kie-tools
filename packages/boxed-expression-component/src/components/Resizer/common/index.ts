/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * Returns a valid width value.
 */
import { DEFAULT_MIN_WIDTH } from "../dom";

export const widthValue = (width: number | string | undefined | null): number => {
  return Math.max(Math.round(parseFloat(width + "")), DEFAULT_MIN_WIDTH);
};

/*
 * Generates a global supervisor hash for a given object.
 */
export const hashfy = (obj = {}): string => {
  const getCircularReplacer = () => {
    const seen = new WeakSet();
    return (key: string, value: unknown) => {
      if (typeof value === "object" && value !== null) {
        if (seen.has(value)) {
          return;
        }
        seen.add(value);
      }
      return value;
    };
  };

  return JSON.stringify(obj, getCircularReplacer());
};

/*
 * Propagate Cell width from DOM to React state.
 */
export const notifyCell = (id: string, width: number = DEFAULT_MIN_WIDTH, editorElement: HTMLElement): void => {
  editorElement.dispatchEvent(
    new CustomEvent(id, {
      detail: { width },
    })
  );
};
