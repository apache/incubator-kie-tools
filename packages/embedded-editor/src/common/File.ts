/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { EditorType } from "./EditorTypes";

/**
 * Definition of a File supported by the embedded-editor.
 */
export interface File {
  fileName: string;
  editorType: EditorType;
  getFileContents: () => Promise<string | undefined>;
  isReadOnly: boolean;
}

/**
 * Default implementation of an empty DMN file; used when creating _new_ DMN files.
 */
export const EMPTY_FILE_DMN = {
  fileName: "new-file",
  editorType: EditorType.DMN,
  getFileContents: () => Promise.resolve(""),
  isReadOnly: false
};

/**
 * Default implementation of an empty BPMN file; used when creating _new_ BPMN files.
 */
export const EMPTY_FILE_BPMN = {
  fileName: "new-file",
  editorType: EditorType.BPMN,
  getFileContents: () => Promise.resolve(""),
  isReadOnly: false
};

/**
 * Default implementation of an empty SCESIM file; used when creating _new_ SCESIM files.
 */
export const EMPTY_FILE_SCESIM = {
  fileName: "new-file",
  editorType: EditorType.SCESIM,
  getFileContents: () => Promise.resolve(""),
  isReadOnly: false
};

/**
 * Helper method to create new, empty files, for different EditorTypes.
 * @param editorType The type of editor for which the new file represents.
 */
export function newFile(editorType: EditorType): File {
  return {
    fileName: "new-file",
    editorType: editorType,
    getFileContents: () => Promise.resolve(""),
    isReadOnly: false
  };
}
