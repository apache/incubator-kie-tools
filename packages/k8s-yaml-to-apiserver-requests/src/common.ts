/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

export type K8sResourceYamlMetadata = {
  name?: string;
  namespace?: string;
};

export type K8sResourceYaml = {
  apiVersion: string;
  kind: string;
  metadata?: K8sResourceYamlMetadata;
};

export type K8sApiServerEndpointByResourceKind = Map<
  string,
  Map<string, { url: { namespaced?: string; global: string }; path: { namespaced?: string; global: string } }>
>;

export function isValidK8sResource(content: any): content is K8sResourceYaml {
  return (
    "apiVersion" in content &&
    typeof content.apiVersion == "string" &&
    "kind" in content &&
    typeof content.kind == "string" &&
    (!("metadata" in content) || ("metadata" in content && typeof content === "object"))
  );
}

export function consoleDebugMessage(message: any) {
  console.debug("K8S YAML TO APISERVER:", message);
}
