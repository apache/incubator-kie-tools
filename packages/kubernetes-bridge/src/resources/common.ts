/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { IObjectMeta } from "@kubernetes-models/apimachinery/apis/meta/v1/ObjectMeta";
import { KubernetesLabelNames } from "./kubernetes/api";
import { OpenShiftLabelNames } from "./openshift/api";

export type IpProtocol = "TCP" | "UDP" | "SCTP";

export type ResourceRequirements = {
  limits: {
    memory: string;
  };
};

export type ResourceMetadata = IObjectMeta & { name: string; namespace: string };

export interface ResourceDescriptor {
  apiVersion: string;
  kind: string;
  metadata: ResourceMetadata;
}

export interface ResourceMetadataEnforcer {
  metadata: ResourceMetadata;
}

export interface ResourceGroupDescriptor<T extends ResourceDescriptor> {
  items: T[];
}

export type Resource<T extends ResourceDescriptor = ResourceDescriptor> =
  | ResourceDescriptor
  | ResourceGroupDescriptor<T>;

export interface EnvVar {
  name: string;
  value: string;
}

export const BUILD_IMAGE_TAG_VERSION = "1.0";

export const ResourceLabelNames = {
  URI: "kogito.kie.org/uri",
  CREATED_BY: "kogito.kie.org/created-by",
  WORKSPACE_NAME: "kogito.kie.org/workspace-name",
};

export const commonLabels = (args: { resourceName: string; createdBy: string }) => ({
  [KubernetesLabelNames.APP]: args.resourceName,
  [KubernetesLabelNames.COMPONENT]: args.resourceName,
  [KubernetesLabelNames.INSTANCE]: args.resourceName,
  [KubernetesLabelNames.PART_OF]: args.resourceName,
  [KubernetesLabelNames.NAME]: args.resourceName,
  [ResourceLabelNames.CREATED_BY]: args.createdBy,
});

export const runtimeLabels = () => ({
  [OpenShiftLabelNames.RUNTIME]: "quarkus",
  [OpenShiftLabelNames.VERSION]: "openjdk-17-el7",
});

export enum DeploymentState {
  UP = "UP",
  DOWN = "DOWN",
  IN_PROGRESS = "IN_PROGRESS",
  PREPARING = "PREPARING",
  ERROR = "ERROR",
}

export interface DeployedModel {
  resourceName: string;
  routeUrl: string;
  creationTimestamp: Date;
  state: DeploymentState;
}

export enum ResourceDataSource {
  PROVIDED = "PROVIDED",
  TEMPLATE = "TEMPLATE",
}
