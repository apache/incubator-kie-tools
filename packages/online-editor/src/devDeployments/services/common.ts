/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

export enum DeploymentState {
  UP = "UP",
  DOWN = "DOWN",
  IN_PROGRESS = "IN_PROGRESS",
  PREPARING = "PREPARING",
  ERROR = "ERROR",
}

export enum HealthStatus {
  UP = "UP",
  ERROR = "ERROR",
}

export interface DeployedModel {
  resourceName: string;
  routeUrl: string;
  creationTimestamp: Date;
  state: DeploymentState;
}
