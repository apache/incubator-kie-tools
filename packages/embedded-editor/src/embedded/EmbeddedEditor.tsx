/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  ChannelType,
  EditorEnvelopeLocator,
  KogitoEditorChannel,
  KogitoEditorChannelApi,
  useConnectedKogitoEditorChannel
} from "@kogito-tooling/microeditor-envelope-protocol";
import { useSyncedKeyboardEvents } from "@kogito-tooling/keyboard-shortcuts-channel";
import { KogitoGuidedTour } from "@kogito-tooling/guided-tour";
import * as CSS from "csstype";
import * as React from "react";
import { useEffect, useImperativeHandle, useMemo, useRef } from "react";
import { File } from "../common";
import { StateControl } from "../stateControl";
import { EditorApi } from "@kogito-tooling/editor-api";
import { KogitoEditorChannelApiImpl } from "./KogitoEditorChannelApiImpl";

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

type ChannelApiMethodsAlreadyImplementedByEmbeddedEditor =
  | "receive_guidedTourUserInteraction"
  | "receive_guidedTourRegisterTutorial"
  | "receive_contentRequest";

type EmbeddedEditorChannelApiOverrides = Partial<
  Omit<KogitoEditorChannelApi, ChannelApiMethodsAlreadyImplementedByEmbeddedEditor>
>;

export type Props = EmbeddedEditorChannelApiOverrides & {
  file: File;
  editorEnvelopeLocator: EditorEnvelopeLocator;
  channelType: ChannelType;
};

/**
 * Forward reference for the `EmbeddedEditor` to support consumers to call upon embedded operations.
 */
export type EmbeddedEditorRef = (EditorApi & { getStateControl(): StateControl }) | null;

const containerStyles: CSS.Properties = {
  display: "flex",
  flex: 1,
  flexDirection: "column",
  width: "100%",
  height: "100%",
  border: "none",
  margin: 0,
  padding: 0,
  overflow: "hidden"
};

const RefForwardingEmbeddedEditor: React.RefForwardingComponent<EmbeddedEditorRef, Props> = (
  props: Props,
  forwardedRef
) => {
  const iframeRef = useRef<HTMLIFrameElement>(null);
  const stateControl = useMemo(() => new StateControl(), []);

  const envelopeMapping = useMemo(() => props.editorEnvelopeLocator.mapping.get(props.file.fileExtension), [
    props.editorEnvelopeLocator,
    props.file
  ]);

  //Setup envelope bus communication
  const kogitoEditorChannelApiImpl = useMemo(() => {
    return new KogitoEditorChannelApiImpl(stateControl, props.file, props);
  }, [stateControl, props.file, props]);

  const kogitoEditorChannel = useMemo(() => {
    return new KogitoEditorChannel({
      postMessage: message => {
        if (iframeRef.current && iframeRef.current.contentWindow) {
          iframeRef.current.contentWindow.postMessage(message, "*");
        }
      }
    });
  }, []);

  useConnectedKogitoEditorChannel(
    kogitoEditorChannel,
    kogitoEditorChannelApiImpl,
    props.editorEnvelopeLocator.targetOrigin,
    { fileExtension: props.file.fileExtension, resourcesPathPrefix: envelopeMapping?.resourcesPathPrefix ?? "" }
  );

  useEffect(() => {
    KogitoGuidedTour.getInstance().registerPositionProvider((selector: string) =>
      kogitoEditorChannel.request_guidedTourElementPositionResponse(selector).then(position => {
        const parentRect = iframeRef.current?.getBoundingClientRect();
        KogitoGuidedTour.getInstance().onPositionReceived(position, parentRect);
      })
    );
  }, [kogitoEditorChannel]);

  // Forward keyboard events to envelope
  useSyncedKeyboardEvents(kogitoEditorChannel.client);

  //Forward reference methods
  useImperativeHandle(
    forwardedRef,
    () => {
      if (!iframeRef.current) {
        return null;
      }

      return {
        getStateControl: () => stateControl,
        getElementPosition: selector => kogitoEditorChannel.request_guidedTourElementPositionResponse(selector),
        redo: () => Promise.resolve(kogitoEditorChannel.notify_editorRedo()),
        undo: () => Promise.resolve(kogitoEditorChannel.notify_editorUndo()),
        getContent: () => kogitoEditorChannel.request_contentResponse().then(c => c.content),
        getPreview: () => kogitoEditorChannel.request_previewResponse(),
        setContent: async content => kogitoEditorChannel.notify_contentChanged({ content: content })
      };
    },
    [kogitoEditorChannel]
  );

  return (
    <>
      {!envelopeMapping && (
        <>
          <span>{`No Editor available for '${props.file.fileExtension}' extension`}</span>
        </>
      )}
      {envelopeMapping && (
        <iframe
          ref={iframeRef}
          id={"kogito-iframe"}
          data-testid={"kogito-iframe"}
          src={envelopeMapping.envelopePath}
          title="Kogito editor"
          style={containerStyles}
          data-envelope-channel={props.channelType}
        />
      )}
    </>
  );
};

export const EmbeddedEditor = React.forwardRef(RefForwardingEmbeddedEditor);
