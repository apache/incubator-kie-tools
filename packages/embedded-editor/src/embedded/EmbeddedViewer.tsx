/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  ChannelType,
  ResourceContent,
  ResourceContentRequest,
  ResourceListRequest,
  ResourcesList
} from "@kogito-tooling/microeditor-envelope-protocol";
import * as React from "react";
import { useCallback } from "react";
import { File } from "../common/File";
import { EmbeddedEditor } from "./EmbeddedEditor";
import { EmbeddedEditorRouter } from "./EmbeddedEditorRouter";

/**
 * Properties supported by the `EmbeddedEditor`.
 */
interface Props {
  /**
   * File to show in the editor.
   */
  file: File;
  /**
   * Router to map editor URLs to installations.
   */
  router: EmbeddedEditorRouter;
  /**
   * Channel in which the editor has been embedded.
   */
  channelType: ChannelType;
  /**
   * Optional callback for when the editor is requesting external content.
   */
  onResourceContentRequest?: (request: ResourceContentRequest) => Promise<ResourceContent | undefined>;
  /**
   * Optional callback for when the editor is requesting a list of external content.
   */
  onResourceListRequest?: (request: ResourceListRequest) => Promise<ResourcesList>;
  /**
   * Optional relative URL for the `envelope.html` used as the inner bus `IFRAME`. Defaults to `envelope/envelope.html`
   */
  envelopeUri?: string;
}

export const EmbeddedViewer = (props: Props) => {
  const noop = useCallback((...args: any) => {
    /*NO OP*/
  }, []);

  const onResourceContentRequest = useCallback(
    (request: ResourceContentRequest) => {
      if (props.onResourceContentRequest) {
        return props.onResourceContentRequest(request);
      }
      return Promise.resolve(new ResourceContent(request.path, undefined));
    },
    [props.onResourceContentRequest]
  );

  const onResourceListRequest = useCallback(
    (request: ResourceListRequest) => {
      if (props.onResourceListRequest) {
        return props.onResourceListRequest(request);
      }
      return Promise.resolve(new ResourcesList(request.pattern, []));
    },
    [props.onResourceListRequest]
  );

  return (
    <EmbeddedEditor
      file={props.file}
      router={props.router}
      channelType={props.channelType}
      onSetContentError={noop}
      onReady={noop}
      onResourceContentRequest={onResourceContentRequest}
      onResourceListRequest={onResourceListRequest}
      onEditorUndo={noop}
      onEditorRedo={noop}
      onOpenFile={noop}
      onNewEdit={noop}
      envelopeUri={props.envelopeUri}
    />
  );
};
