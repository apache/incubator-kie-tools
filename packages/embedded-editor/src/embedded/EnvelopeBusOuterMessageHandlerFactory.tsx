/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { EditorContent, KogitoEdit, ResourceContent, ResourceContentRequest, ResourceListRequest, ResourcesList } from "@kogito-tooling/core-api";
import { EnvelopeBusOuterMessageHandler } from "@kogito-tooling/microeditor-envelope-protocol";
import * as React from "react";
import { Props } from "./EmbeddedEditor";

export function newEnvelopeBusOuterMessageHandler(
  props: Props,
  iframeRef: React.RefObject<HTMLIFrameElement>,
  onContentResponse: (content: EditorContent) => void,
  onSetContentError: (errorMessage: string) => void,
  onDirtyIndicatorChange: (isDirty: boolean) => void,
  onReady: () => void,
  onResourceContentRequest: (request: ResourceContentRequest) => Promise<ResourceContent | undefined>,
  onResourceListRequest: (request: ResourceListRequest) => Promise<ResourcesList>,
  onEditorUndo: () => void,
  onEditorRedo: () => void,
  onNewEdit: (edit: KogitoEdit) => void,
  onPreviewResponse: (previewSvg: string) => void
): EnvelopeBusOuterMessageHandler {
  return new EnvelopeBusOuterMessageHandler(
    {
      postMessage: msg => {
        if (iframeRef.current && iframeRef.current.contentWindow) {
          iframeRef.current.contentWindow.postMessage(msg, "*");
        }
      }
    },
    self => ({
      pollInit() {
        self.request_initResponse(window.location.origin);
      },
      receive_languageRequest() {
        self.respond_languageRequest(props.router.getLanguageData(props.file.editorType));
      },
      receive_contentResponse(content: EditorContent) {
        onContentResponse(content);
      },
      receive_contentRequest() {
        props.file
          .getFileContents()
          .then(c => self.respond_contentRequest({ content: c ?? "", path: props.file.fileName }));
      },
      receive_setContentError(errorMessage: string) {
        onSetContentError(errorMessage);
      },
      receive_dirtyIndicatorChange(isDirty: boolean) {
        onDirtyIndicatorChange(isDirty);
      },
      receive_ready() {
        onReady();
      },
      receive_resourceContentRequest(request: ResourceContentRequest) {
        onResourceContentRequest(request).then(r => self.respond_resourceContent(r!));
      },
      receive_resourceListRequest(request: ResourceListRequest) {
        onResourceListRequest(request).then(r => self.respond_resourceList(r!));
      },
      notify_editorUndo: () => {
        onEditorUndo();
      },
      notify_editorRedo: () => {
        onEditorRedo();
      },
      receive_newEdit(edit: KogitoEdit) {
        onNewEdit(edit);
      },
      receive_previewRequest(previewSvg: string) {
        onPreviewResponse(previewSvg);
      }
    }));
}

