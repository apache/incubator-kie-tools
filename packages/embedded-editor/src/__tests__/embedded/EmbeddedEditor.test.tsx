/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  ChannelEnvelopeServer,
  ChannelType,
  EditorEnvelopeLocator,
  KogitoEdit,
  KogitoEditorChannelEnvelopeServer
} from "@kogito-tooling/editor-envelope-protocol";
import * as React from "react";
import { File } from "../../common";
import { EmbeddedEditor, EmbeddedEditorRef } from "../../embedded";
import { incomingMessage } from "./EmbeddedEditorTestUtils";
import { render } from "@testing-library/react";
import { EnvelopeBusMessagePurpose } from "@kogito-tooling/envelope-bus";

describe("EmbeddedEditor::ONLINE", () => {
  const file: File = {
    fileName: "test",
    fileExtension: "dmn",
    getFileContents: () => Promise.resolve(""),
    isReadOnly: false
  };

  const editorEnvelopeLocator: EditorEnvelopeLocator = {
    targetOrigin: "localhost:8888",
    mapping: new Map([["dmn", { envelopePath: "envelope/envelope.html", resourcesPathPrefix: "envelope" }]])
  };

  const channelType = ChannelType.ONLINE;
  const editorRef = React.createRef<EmbeddedEditorRef>();
  const busId = "test-bus-id";

  beforeEach(() => {
    jest.spyOn(ChannelEnvelopeServer.prototype, "generateRandomId").mockReturnValue(busId);
    jest.clearAllMocks();
  });

  test("EmbeddedEditor::defaults", () => {
    const { getByTestId, container } = render(
      <EmbeddedEditor
        ref={editorRef}
        file={file}
        editorEnvelopeLocator={editorEnvelopeLocator}
        channelType={channelType}
      />
    );

    expect(getByTestId("kogito-iframe")).toBeVisible();
    expect(getByTestId("kogito-iframe")).toHaveAttribute("data-envelope-channel", ChannelType.ONLINE);
    expect(getByTestId("kogito-iframe")).toHaveAttribute("src", "envelope/envelope.html");

    expect(container.firstChild).toMatchSnapshot();
  });

  test("EmbeddedEditor::setContent", () => {
    const spyRespond_contentRequest = jest.spyOn(KogitoEditorChannelEnvelopeServer.prototype, "notify_contentChanged");
    render(
      <EmbeddedEditor
        ref={editorRef}
        file={file}
        editorEnvelopeLocator={editorEnvelopeLocator}
        channelType={channelType}
      />
    );
    editorRef.current?.setContent("content", "");

    expect(spyRespond_contentRequest).toBeCalledWith({ content: "content" });
  });

  test("EmbeddedEditor::requestContent", () => {
    const spyRequest_contentResponse = jest.spyOn(KogitoEditorChannelEnvelopeServer.prototype, "request_contentResponse");
    render(
      <EmbeddedEditor
        ref={editorRef}
        file={file}
        editorEnvelopeLocator={editorEnvelopeLocator}
        channelType={channelType}
      />
    );
    editorRef.current?.getContent();

    expect(spyRequest_contentResponse).toBeCalled();
  });

  test("EmbeddedEditor::requestPreview", () => {
    const spyRequest_previewResponse = jest.spyOn(KogitoEditorChannelEnvelopeServer.prototype, "request_previewResponse");
    render(
      <EmbeddedEditor
        ref={editorRef}
        file={file}
        editorEnvelopeLocator={editorEnvelopeLocator}
        channelType={channelType}
      />
    );
    editorRef.current?.getPreview();

    expect(spyRequest_previewResponse).toBeCalled();
  });

  test("EmbeddedEditor::onSetContentError", async () => {
    const onSetContentError = jest.fn();

    const { container } = render(
      <EmbeddedEditor
        ref={editorRef}
        file={file}
        editorEnvelopeLocator={editorEnvelopeLocator}
        channelType={channelType}
        receive_setContentError={onSetContentError}
      />
    );

    await incomingMessage({
      busId: busId,
      purpose: EnvelopeBusMessagePurpose.NOTIFICATION,
      type: "receive_setContentError",
      data: []
    });

    expect(onSetContentError).toBeCalled();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("EmbeddedEditor::onReady", async () => {
    const onReady = jest.fn();

    const { container } = render(
      <EmbeddedEditor
        ref={editorRef}
        file={file}
        editorEnvelopeLocator={editorEnvelopeLocator}
        channelType={channelType}
        receive_ready={onReady}
      />
    );

    await incomingMessage({
      busId: busId,
      purpose: EnvelopeBusMessagePurpose.NOTIFICATION,
      type: "receive_ready",
      data: []
    });

    expect(onReady).toBeCalled();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("EmbeddedEditor::onResourceContentRequest", async () => {
    const onResourceContentRequest = jest.fn(() => Promise.resolve({} as any));

    const { container } = render(
      <EmbeddedEditor
        ref={editorRef}
        file={file}
        editorEnvelopeLocator={editorEnvelopeLocator}
        channelType={channelType}
        receive_resourceContentRequest={onResourceContentRequest}
      />
    );

    await incomingMessage({
      busId: busId,
      requestId: "1",
      purpose: EnvelopeBusMessagePurpose.REQUEST,
      type: "receive_resourceContentRequest",
      data: [{ path: "" }]
    });

    expect(onResourceContentRequest).toBeCalled();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("EmbeddedEditor::onResourceListRequest", async () => {
    const onResourceListRequest = jest.fn(() => Promise.resolve({} as any));

    const { container } = render(
      <EmbeddedEditor
        ref={editorRef}
        file={file}
        editorEnvelopeLocator={editorEnvelopeLocator}
        channelType={channelType}
        receive_resourceListRequest={onResourceListRequest}
      />
    );

    await incomingMessage({
      busId: busId,
      requestId: "1",
      purpose: EnvelopeBusMessagePurpose.REQUEST,
      type: "receive_resourceListRequest",
      data: [{ pattern: "", paths: [] }]
    });

    expect(onResourceListRequest).toBeCalled();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("EmbeddedEditor::onNewEdit", async () => {
    const onNewEdit = jest.fn();

    const { container } = render(
      <EmbeddedEditor
        ref={editorRef}
        file={file}
        editorEnvelopeLocator={editorEnvelopeLocator}
        channelType={channelType}
        receive_newEdit={onNewEdit}
      />
    );

    await incomingMessage({
      busId: busId,
      purpose: EnvelopeBusMessagePurpose.NOTIFICATION,
      type: "receive_newEdit",
      data: [new KogitoEdit("1")]
    });

    expect(editorRef.current?.getStateControl().getCommandStack()).toEqual(["1"]);
    expect(onNewEdit).toBeCalled();
    expect(container.firstChild).toMatchSnapshot();
  });
});
