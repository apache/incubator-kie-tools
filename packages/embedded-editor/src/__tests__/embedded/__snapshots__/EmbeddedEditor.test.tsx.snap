// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmbeddedEditor::ONLINE EmbeddedEditor::defaults 1`] = `
<body>
  <div>
    <iframe
      data-envelope-channel="ONLINE"
      data-testid="kogito-iframe"
      id="kogito-iframe"
      src="envelope/envelope.html"
      style="display: flex; flex: 1; flex-direction: column; width: 100%; height: 100%; margin: 0px; padding: 0px; overflow: hidden;"
      title="Kogito editor"
    />
  </div>
</body>
`;

exports[`EmbeddedEditor::ONLINE EmbeddedEditor::onContentResponse 1`] = `
<body>
  <div>
    <iframe
      data-envelope-channel="ONLINE"
      data-testid="kogito-iframe"
      id="kogito-iframe"
      src="envelope/envelope.html"
      style="display: flex; flex: 1; flex-direction: column; width: 100%; height: 100%; margin: 0px; padding: 0px; overflow: hidden;"
      title="Kogito editor"
    />
  </div>
</body>
`;

exports[`EmbeddedEditor::ONLINE EmbeddedEditor::onDirtyIndicatorChange 1`] = `
<body>
  <div>
    <iframe
      data-envelope-channel="ONLINE"
      data-testid="kogito-iframe"
      id="kogito-iframe"
      src="envelope/envelope.html"
      style="display: flex; flex: 1; flex-direction: column; width: 100%; height: 100%; margin: 0px; padding: 0px; overflow: hidden;"
      title="Kogito editor"
    />
  </div>
</body>
`;

exports[`EmbeddedEditor::ONLINE EmbeddedEditor::onNewEdit 1`] = `
<body>
  <div>
    <iframe
      data-envelope-channel="ONLINE"
      data-testid="kogito-iframe"
      id="kogito-iframe"
      src="envelope/envelope.html"
      style="display: flex; flex: 1; flex-direction: column; width: 100%; height: 100%; margin: 0px; padding: 0px; overflow: hidden;"
      title="Kogito editor"
    />
  </div>
</body>
`;

exports[`EmbeddedEditor::ONLINE EmbeddedEditor::onPreviewResponse 1`] = `
<body>
  <div>
    <iframe
      data-envelope-channel="ONLINE"
      data-testid="kogito-iframe"
      id="kogito-iframe"
      src="envelope/envelope.html"
      style="display: flex; flex: 1; flex-direction: column; width: 100%; height: 100%; margin: 0px; padding: 0px; overflow: hidden;"
      title="Kogito editor"
    />
  </div>
</body>
`;

exports[`EmbeddedEditor::ONLINE EmbeddedEditor::onReady 1`] = `
<body>
  <div>
    <iframe
      data-envelope-channel="ONLINE"
      data-testid="kogito-iframe"
      id="kogito-iframe"
      src="envelope/envelope.html"
      style="display: flex; flex: 1; flex-direction: column; width: 100%; height: 100%; margin: 0px; padding: 0px; overflow: hidden;"
      title="Kogito editor"
    />
  </div>
</body>
`;

exports[`EmbeddedEditor::ONLINE EmbeddedEditor::onResourceContentRequest 1`] = `
<body>
  <div>
    <iframe
      data-envelope-channel="ONLINE"
      data-testid="kogito-iframe"
      id="kogito-iframe"
      src="envelope/envelope.html"
      style="display: flex; flex: 1; flex-direction: column; width: 100%; height: 100%; margin: 0px; padding: 0px; overflow: hidden;"
      title="Kogito editor"
    />
  </div>
</body>
`;

exports[`EmbeddedEditor::ONLINE EmbeddedEditor::onResourceListRequest 1`] = `
<body>
  <div>
    <iframe
      data-envelope-channel="ONLINE"
      data-testid="kogito-iframe"
      id="kogito-iframe"
      src="envelope/envelope.html"
      style="display: flex; flex: 1; flex-direction: column; width: 100%; height: 100%; margin: 0px; padding: 0px; overflow: hidden;"
      title="Kogito editor"
    />
  </div>
</body>
`;

exports[`EmbeddedEditor::ONLINE EmbeddedEditor::onSetContentError 1`] = `
<body>
  <div>
    <iframe
      data-envelope-channel="ONLINE"
      data-testid="kogito-iframe"
      id="kogito-iframe"
      src="envelope/envelope.html"
      style="display: flex; flex: 1; flex-direction: column; width: 100%; height: 100%; margin: 0px; padding: 0px; overflow: hidden;"
      title="Kogito editor"
    />
  </div>
</body>
`;
